source/array_list.c

```diff
@@ -260,3 +260,40 @@ int aws_array_list_set_at(struct aws_array_list *list, const void *val, size_t i
     return AWS_OP_SUCCESS;
 }
 
+
+static void aws_array_list_mem_swap(void * restrict item1, void * restrict item2, size_t item_size) {
+#define SLICE 128
+
+    /* copy SLICE sized bytes at a time */
+    size_t slice_count = item_size / SLICE;
+    uint8_t temp[SLICE];
+    for(size_t i = 0; i < slice_count; i++) {
+        memcpy(temp, item1, SLICE);
+        memcpy(item1, item2, SLICE);
+        memcpy(item2, temp, SLICE);
+        item1 = (uint8_t*)item1 + SLICE;
+        item2 = (uint8_t*)item2 + SLICE;
+    }
+
+    size_t remainder = item_size % SLICE;
```

Comment:
size_t remainder = item_size & 0x7F

is likely faster, and this is a hot path, so it would probably be better.

Reply:
or better yet, 

size_t remainder = item_size & (SLICE - 1);

Reply:
also clang and msvc love to complain about these not being cast to (void *)

Reply:
Normally compilers can optimize `%` on powers of two, so in general I don't worry about this too much (unless it's not actually a power of two).